pkg/pool-stable/contracts/ComposableStablePoolProtocolFees.sol

```diff
@@ -227,44 +232,68 @@ abstract contract ComposableStablePoolProtocolFees is
         // Note that the pre-join/exit invariant is *not* the invariant from the last join,
         // but computed from the balances before this particular join/exit.
 
-        uint256 protocolSwapFeePercentage = getProtocolFeePercentageCache(ProtocolFeeType.SWAP);
-
-        if (protocolSwapFeePercentage > 0) {
-            uint256 invariantGrowthRatio = (
-                postJoinExitInvariant > preJoinExitInvariant
-                    ? postJoinExitInvariant.sub(preJoinExitInvariant)
-                    : preJoinExitInvariant.sub(postJoinExitInvariant)
-            )
-                .divDown(preJoinExitInvariant);
-
-            // Compute the bpt ratio
-            uint256 bptGrowthRatio = (
-                postJoinExitSupply > preJoinExitInvariant
-                    ? postJoinExitSupply.sub(preJoinExitSupply)
-                    : preJoinExitSupply.sub(postJoinExitSupply)
-            )
-                .divDown(preJoinExitSupply);
-
-            // The difference between the invariant growth and bpt increase rates must be due to the
-            // balance change from this join/exit.
-            // Protocol fees due = (invariant growth / bpt increase - 1) * virtual supply * protocol fee %
-            // For instance, if the invariant growth is 1.05, and the bpt increase is 1.0475, with 1000 supply,
-            // and a protocol fee of 50%, we would mint (1.05/1.0475 - 1) * 1000 * 0.5 = 1.193 BPT.
-
-            if (invariantGrowthRatio > bptGrowthRatio) {
-                uint256 protocolFeeAmount = invariantGrowthRatio
-                    .divDown(bptGrowthRatio)
-                    .sub(FixedPoint.ONE)
-                    .mulDown(preJoinExitSupply)
-                    .mulDown(protocolSwapFeePercentage);
-
-                _payProtocolFees(protocolFeeAmount);
-            }
+        // `_payProtocolFeesBeforeJoinExit` paid protocol fees accumulated between the previous and current
+        // join or exit, while this code pays any protocol fees due on the current join or exit.
+        // The amp and rates are constant during a single transaction, so it doesn't matter if there
+        // is an ongoing amp change, and we can ignore yield.
+
+        // This represents the total value exchanged in this particular join-exit, including swap fees.
+        //
+        // Note that the "growth" is actually negative in the case of exits: both the supply and invariant
+        // decrease. Since swap fees are symmetrical, it is numerically equivalent (and conceptually simpler)
+        // to just invert the sign in the exit case, and use the same "growth" terminology and calculations
+        // for both.
+        //
+        uint256 invariantGrowth = (
+            postJoinExitInvariant > preJoinExitInvariant
+                ? postJoinExitInvariant - preJoinExitInvariant
+                : preJoinExitInvariant - postJoinExitInvariant
+        )
+            .divDown(preJoinExitInvariant);
```

Comment:
I find this slightly concerning, because due to rounding errors we might have post join invariants that are actually smaller than the pre join. Could we add a boolean to this function that indicates whether to expect a join or an exit, and then do the expected subtraction depending on it (capping to 0 if the result would be negative)?

Reply:
Addressed in [1ce7bf8](https://github.com/balancer-labs/balancer-v2-monorepo/pull/1603/commits/1ce7bf84a301c534494cb453fc1651c6adaa3ee6).